README.md
# Visual Studio Code - Discord RPC
## To have a language added, submit a pull request adding to the supported languages list and add a 512x512 png image with the language's icon.

CHANGELOG.md
# Change Log
All notable changes to the "vsc-discord-rpc" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]
- Initial release

extension.js
const { Client } = require("discord-rpc");
const { extname, basename } = require("path");
const { workspace, commands, window, StatusBarAlignment } = require("vscode");

const languages = require("./languages");

let client, config;

function activate(context) {
    config = workspace.getConfiguration("discordrp");

    if (config.get("enabled")) client = new RPC(config.get("clientID"));

    context.subscriptions.push(
        commands.registerCommand("discordrp.enable", () => {
            if (client) client.destroy();

            config.update("enabled", true);

            client = new RPC(config.get("clientID"));

            window.showInformationMessage("Discord Rich Presense is now enabled.");
        }),
        commands.registerCommand("discordrp.disable", () => {
            if (!client) return;

            config.update("enabled", false);

            client.destroy();

            window.showInformationMessage("Discord Rich Presense is now disabled.");
        })
    );
}

function deactivate(context) {
    if (client) client.destroy();
}

exports.activate = activate;
exports.deactivate = deactivate;

class RPC extends Client {
    constructor(clientID) {
        super({ transport: "ipc" });

        this.eventHandler;
        this.statusBarItem;

        this.reconnect = null;
        this.reconnectionAttempts = 0;

        this.lastFile = null;

        this.once("ready", () => {
            if (this.reconnect) { clearInterval(this.reconnect); this.reconnect = null; }

            this.reconnectionAttempts = 0;

            this.setActivity();
            
            this.eventHandler = workspace.onDidChangeTextDocument(() => this.setActivity());

            this.statusBarItem = window.createStatusBarItem(StatusBarAlignment.Left);

            this.statusBarItem.text = config.get("statusBarText");
            this.statusBarItem.show();
            
            this.transport.once("close", () => {
                if (!config.get("enabled")) return;

                this.destroy();

                this.reconnect = setInterval(() => {
                    this.reconnectionAttempts++;

                    client = new RPC(config.get("clientID"));
                }, 5000);
            });
        });

        this.login(clientID).catch(err => {
            if (this.reconnect && this.reconnectionAttempts >= config.get("reconnectThreshold")) this.destroy();

            if (err.message.includes("ENOENT")) return window.showErrorMessage("A Discord client cannot be detected.");
            window.showErrorMessage(`An error occured while trying to connected to Discord via RPC: ${err.message}`);
        });
    }

    destroy() {
        client = null;
        
        if (this.reconnect) clearInterval(this.reconnect);
        
        this.clearActivity();

        this.eventHandler.dispose();

        this.statusBarItem.dispose();

        super.destroy();
    }

    setActivity() {
        if (window.activeTextEditor && window.activeTextEditor.document.fileName === this.lastFile) return;

        let data;

        if (window.activeTextEditor) {
            this.lastFile = window.activeTextEditor.document.fileName;

            const details = config.get("details").replace("{file}", basename(window.activeTextEditor.document.fileName));
            const checkState = !!workspace.getWorkspaceFolder(window.activeTextEditor.document.uri);
            const state = checkState ? config.get("workspace").replace("{workspace}", workspace.getWorkspaceFolder(window.activeTextEditor.document.uri).name) : config.get("workspaceNotFound");

            const ext = extname(basename(window.activeTextEditor.document.fileName)).substring(1) || basename(window.activeTextEditor.document.fileName).substring(1);
            const lang = languages[ext] || { "title": ext, "key": "file" };
         
            data = {
                details,
                state,
                startTimestamp: Date.now() / 1000,
                largeImageKey: lang.key,
                largeImageText: config.get("largeImage").replace("{language}", lang.title),
                smallImageKey: "vsc",
                smallImageText: config.get("smallImage"),
                instance: false
            };
        } else {
            const details = config.get("detailsIdle");
            const checkState = false;
            const state = config.get("workspaceIdle");

            data = {
                details,
                state,
                startTimestamp: Date.now() / 1000,
                largeImageKey: "vsc-large",
                largeImageText: config.get("largeImageIdle"),
                smallImageKey: "vsc",
                smallImageText: config.get("smallImage"),
                instance: false
            };
        }
        
        super.setActivity(data);
    }

    clearActivity() {
        super.setActivity({});
    }
}


old-extension.js
const discord_rpc = require("discord-rpc");
const path = require("path");
const timers = require("timers");
const vscode = require("vscode");

let rpc, eventHandler, config, reconnect, reconnectCounter = 0, lastKnownFileName;

function activate(context) {
    config = vscode.workspace.getConfiguration('discord');
    
    if (config.get('enabled')) initRPC(config.get('clientID'));

    const enabler = vscode.commands.registerCommand('discord.enable', () => {
        if (rpc) destroyRPC();

        config.update('enabled', true);

        initRPC(config.get('clientID'));
        
        vscode.window.showInformationMessage('Enabled Discord Rich Presence for this workspace.');
    });
    
    const disabler = vscode.commands.registerCommand('discord.disable', () => {
        if (!rpc) return;

        config.update('enabled', false);

        rpc.setActivity({});

        destroyRPC();
        
        vscode.window.showInformationMessage('Disabled Discord Rich Presence for this workspace.');
    });
    
    context.subscriptions.push(enabler, disabler);
}

function deactivate(context) { destroyRPC(); }

function initRPC(clientID) {
    rpc = new discord_rpc.Client({ transport: 'ipc' });
    
    rpc.once('ready', () => {
        if (reconnect) {
            timers.clearInterval(reconnect);
            reconnect = null;
        }
        
        reconnectCounter = 0;

        setActivity();

        eventHandler = vscode.workspace.onDidChangeTextDocument((e) => setActivity());
        
        rpc.transport.once('close', () => {
            if (!config.get('enabled')) return;
           
            destroyRPC();
            
            reconnect = timers.setInterval(() => {
                reconnectCounter++;

                initRPC(config.get('clientID'));
            }, 5000);
        });
    });
    
    rpc.login(clientID).catch(error => {
        if (reconnect) {
            if (reconnectCounter >= config.get('reconnectThreshold')) destroyRPC();
            else return;
        }

        if (error.message.includes('ENOENT')) vscode.window.showErrorMessage('No Discord Client detected!');
        else vscode.window.showErrorMessage(`Couldn't connect to discord via rpc: ${error.message}`);
    });
}

function destroyRPC() {
    if (!rpc) return;
    
    if (reconnect) timers.clearInterval(reconnect);
    
    reconnect = null;
    
    eventHandler.dispose();
    
    rpc.destroy();
    
    rpc = null;
    
    lastKnownFileName = null;
}

function language(id, check = false) {
    const languages = { "javascript": "JavaScript", "html": "HTML", "css": "CSS", "json": "JSON" };

    return check ? !!languages[id] : (languages[id] ? `Working in ${languages[id]}` : "Unsupported Language");
}

function hasLanguage(ext) { return ["js", "html", "css", "json", "ejs"].includes(ext); }

function setActivity() {
    if (!rpc) return;
    if (vscode.window.activeTextEditor && vscode.window.activeTextEditor.document.fileName === lastKnownFileName) return;

    lastKnownFileName = vscode.window.activeTextEditor ? vscode.window.activeTextEditor.document.fileName : null;

    const details = vscode.window.activeTextEditor
        ? config.get('details').replace('{filename}', path.basename(vscode.window.activeTextEditor.document.fileName))
        : config.get('detailsIdle');
    const checkState = vscode.window.activeTextEditor
        ? Boolean(vscode.workspace.getWorkspaceFolder(vscode.window.activeTextEditor.document.uri))
        : false;
    const state = vscode.window.activeTextEditor
        ? checkState
            ? config.get('workspace').replace('{workspace}', vscode.workspace.getWorkspaceFolder(vscode.window.activeTextEditor.document.uri).name)
            : config.get('workspaceNotFound')
        : config.get('workspaceIdle');
    
    const largeImageText = vscode.window.activeTextEditor ? config.get('largeImage') || language(vscode.window.activeTextEditor.document.languageId) : config.get('largeImageIdle');
   
    const ext = vscode.window.activeTextEditor ? path.extname(path.basename(vscode.window.activeTextEditor.document.fileName)).substring(1) || path.basename(vscode.window.activeTextEditor.document.fileName).substring(1) : null;
    const largeImageKey = vscode.window.activeTextEditor ? hasLanguage(ext) ? ext : "file" : "vsc-large";

    const activity = {
        details,
        state,
        startTimestamp: Date.now() / 1000,
        largeImageKey,
        largeImageText,
        smallImageKey: 'vsc',
        smallImageText: config.get('smallImage'),
        instance: false
    };
    
    rpc.setActivity(activity);
}

exports.activate = activate;
exports.deactivate = deactivate;